File: Week_04/search_a_2d_matrix/Solution.java
package search_a_2d_matrix;

/**
 * 二分查找
 * 把二维数组视为一维数组
 * Time：O(log mn)
 */
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = (matrix != null) ? matrix.length : 0;
        int n = (matrix.length > 0) ? matrix[0].length : 0;

        int left = 0, right = m * n - 1;
        while (left <= right) {
            int mid = (left + right) >>> 1;
            int num = matrix[mid / n][mid % n];
            if (num == target)
                return true;
            else if (num < target)
                left = mid + 1;
            else
                right = mid - 1;
        }
        return false;
    }
}

File: Week_04/search_a_2d_matrix/Solution.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0]) if m > 0 else 0

        lo, hi = 0, m * n - 1
        while lo <= hi:
            mid = int((lo + hi) / 2)
            val = matrix[int(mid / n)][mid % n]
            if val == target:
                return True
            if val < target:
                lo = mid + 1
            else:
                hi = mid - 1
        return False